### PR TITLE
Support setting the Pull up/Pull down resistor mode on the Pi4

### PR DESCRIPTION
**Pull request by pgrawehr**
The Pi4's CPU uses alternate registers to set the pull up/pull down mode than the Pi3 did. This adds a separate driver class for the Pi4 with (for now) one extra method to configure the registers for setting the pull mode of input registers. The required code was adapted from https://github.com/joan2937/pigpio/blob/master/pigpio.c#L8785. 

Unit test provided as well. 

The detection of the Pi4 is quite a hack, this should be improved later (see also PR #766). It would probably be better to either create a separate class for detecting the best driver or to have the drivers detect themselves whether they can run on the given hardware.

### CODE REVIEW

File: src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            var bestDriver = GpioController.GetBestDriverForBoard();
+            if (!(bestDriver is RaspberryPi4Driver))
+            {
+                bestDriver.Dispose();
+                return;
+            }
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), bestDriver))
+            {
+                controller.OpenPin(OpenPin, PinMode.InputPullUp);
+                Thread.Sleep(10);
```
Comment (MarkCiliaVincenti) [inline src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs]: Can you please explain why you are pausing for 10ms?

**Comment by Frankenslag** [comment]
@pgrawehr 

Some initial thoughts

Are you saying that the pi4 isn’t working with regards to seeing the pull-ups ?

Anyway I think that this or suffers the same issue as #766 tries to solve. Because the hardware and model bits of /proc/cpuinfo are Raspbian and so debian and Ubuntu wouldn’t have the info that you’re looking for. One way might be to look for the value of peripheral base address from the device tree which seems to be there in all cases I have looked at. This value from memory would be 0xFE000000 for the Pi 4.

I do think you’re right about having some cook way for these detections. How about as a static method of the RaspberryPi3Device.

Would this work better are a modification of the standard pi device rather than having a new device?

File: src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs (hunk 1 above)
Comment (pgrawehr) [reply src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs]: Not really. I was first expecting it to take a moment to settle, but (as the remainder of the test shows), the change happens instantly. So this is not needed.

**Comment by pgrawehr** [comment]
> 
> 
> @pgrawehr
> 
> Some initial thoughts
> 
> Are you saying that the pi4 isn’t working with regards to seeing the pull-ups ?

Yes. When attempting to set the pull up mode, the existing Pi3 driver just doesn't change the mode from whatever it is now to the new mode. After some investigations, I found that this requires writing different hardware registers than previously. Also other low-level GPIO implementations (such as wiringpi) have the same problem, some have been updated meanwhile. See my question [here](https://raspberrypi.stackexchange.com/questions/104312/setting-pull-up-pull-down-mode-doesnt-work-mostly) for a confirmation of my observation and the hint to the correct solution (which I implemented now).  

> 
> Anyway I think that this or suffers the same issue as #766 tries to solve. Because the hardware and model bits of /proc/cpuinfo are Raspbian and so debian and Ubuntu wouldn’t have the info that you’re looking for. One way might be to look for the value of peripheral base address from the device tree which seems to be there in all cases I have looked at. This value from memory would be 0xFE000000 for the Pi 4.
I agree, but this should be solved in one place, therefore I just added a hack for this PR. 
> 
> I do think you’re right about having some cook way for these detections. How about as a static method of the RaspberryPi3Device.
Or even as a normal method. One would then create all the devices in order, and each of them would then magically detect whether they work on the current platform (+ eventually whether they're the optimal match for that platform). Now one can manually create all Device implementations, but whether they're any good is just left to the caller. 
> 
> Would this work better are a modification of the standard pi device rather than having a new device?
Not sure about this. For this single change maybe, but I don't know whether there's more to come.

File: src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs [hunk 2]
```diff
@@ -88,4 +88,10 @@ internal unsafe struct RegisterView
     ///<summary>GPIO Pin Pull-up/down Enable Clock, 2x32 bits, R/W.</summary>
     [FieldOffset(0x98)]
     public fixed uint GPPUDCLK[2];
+
+    /// <summary>
+    /// Pull-up/down Enable on the BCM2711, 4x32 bits, R/W (2 bits per Pin)
```
Comment (krwq) [inline src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs]: please use one liner summary similarly as other fields

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.Linux.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.Linux.cs]: copyright header here and any other files it might be missing

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Device.Gpio.Drivers
+{
+    public unsafe partial class RaspberryPi4Driver : RaspberryPi3Driver
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: is this file actually needed? Not sure what's the value in having empty class

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: I mean we can leave it for compat but then I'd change RaspberryPi3Driver to be the same (and not partial) and create GenericRaspberryPiDriver or something similar (or ..Pi3Or4Driver)

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: You mean rename the current RaspberryPi3Driver to GenericRaspberryPiDriver and then create 2 derived classes?

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: Done.

**Comment by pgrawehr** [comment]
This PR is ready for a final review, I think.

**Comment by pgrawehr** [comment]
@Frankenslag : When will PR #766 be ready? I guess it would be easier to combine that directly, otherwise we'll get merge conflicts anyway.

**Comment by Frankenslag** [comment]
> @Frankenslag : When will PR #766 be ready? I guess it would be easier to combine that directly, otherwise we'll get merge conflicts anyway.

@pgrawehr 
Well I have no further work to do on this at the moment I am not sure why it hasn't yet been reviewed and merged. Do you want to combine with your work ? If so let me know and I'll cancel the PR.

**Comment by pgrawehr** [comment]
Leave it open for now, if it gets merged I can rebase to that; Otherwise I'll merge that on my branch, but I'll have to see whether I get to that (I have some other open ends).

File: src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            return new RaspberryPi4Driver();
+        }
+
+        protected override PinNumberingScheme GetTestNumberingScheme() => PinNumberingScheme.Logical;
+
+        [Fact]
+        public void InputPullResistorsWork()
+        {
+            // This is a hack to test whether we're running on a Pi4 without duplicating that detection logic. 
+            var bestDriver = GpioController.GetBestDriverForBoard();
```
Comment (krwq) [inline src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs]: I'm slightly skeptical about this check. If our detection logic is broken this will pass because it will not test anything. Can you at least make it so that test is skipped instead so it's more likely to be spotted?

In corefx we have [ConditionalFact](https://github.com/dotnet/corefx/search?q=ConditionalFact&unscoped_q=ConditionalFact) and [SkipTestException](https://github.com/dotnet/corefx/search?q=SkipTestException&unscoped_q=SkipTestException) for that but you can also use xunit Trait if not sure how to use that (i.e. `[Trait("SkipOnCi", "true")]`). For traits you will need to pass something similar to let our CI know to not run it, so i.e. https://github.com/dotnet/iot/blob/master/eng/sendToHelix.proj#L48 except do it unconditionally

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Device.Gpio.Drivers
+{
+    public unsafe partial class RaspberryPi4Driver : RaspberryPi3Driver
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: should this inherit from GenericRaspberryPiDriver?

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs [hunk 7]
```diff
@@ -28,19 +28,37 @@ public GpioController(PinNumberingScheme numberingScheme)
...
+                            {
+                                string revisionLine = cpuInfoLines[i + 1];
+                                match = regexRev.Match(revisionLine);
+                                if (match.Success && match.Groups.Count > 1)
+                                {
+                                    string rev = match.Groups[1].Value;
+                                    if (rev.EndsWith("03111"))
+                                    {
+                                        // The Pi4 reports the same CPU (although that's technically wrong) but different board revisions.
+                                        // The currently known revisions are 0xa03111, 0xb0311, 0xc03111 for the models with 1GB, 2GB and 4GB of ram, respectively.
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: Do you mean this won't work for 2GB or should this be `0xb03111`? Are the revisions supposed to be the same here?

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs [hunk 8]
```diff
@@ -28,19 +28,37 @@ public GpioController(PinNumberingScheme numberingScheme)
...
+                                    string rev = match.Groups[1].Value;
+                                    if (rev.EndsWith("03111"))
+                                    {
+                                        // The Pi4 reports the same CPU (although that's technically wrong) but different board revisions.
+                                        // The currently known revisions are 0xa03111, 0xb0311, 0xc03111 for the models with 1GB, 2GB and 4GB of ram, respectively. 
+                                        return new RaspberryPi4Driver();
+                                    }
+                                }
+                            }
+                            // Assume Pi3 otherwise
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: nit: enter after `}`

**Comment by krwq** [comment]
Hey, sorry for the delay, this mostly looks good to me but @joperezr is more familiar with this code so he'll need to review this as well

File: src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs [hunk 9]
```diff
@@ -88,4 +88,8 @@ internal unsafe struct RegisterView
     ///<summary>GPIO Pin Pull-up/down Enable Clock, 2x32 bits, R/W.</summary>
     [FieldOffset(0x98)]
     public fixed uint GPPUDCLK[2];
+
+    /// <summary>Pull-up/down Enable on the BCM2711, 4x32 bits, R/W (2 bits per Pin)</summary>
+    [FieldOffset(0xE4)]
+    public fixed uint GPPUPPDN[4];
```
Comment (joperezr) [inline src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs]: Is this available on other Pis or only on Pi4? Also, do you need access to '/dev/mem' memory mapped file or is this register available via '/dev/gpiomem'?

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs [hunk 10]
```diff
@@ -28,19 +28,37 @@ public GpioController(PinNumberingScheme numberingScheme)
         /// Attempt to get the best applicable driver for the board the program is executing on.
         /// </summary>
         /// <returns>A driver that works with the board the program is executing on.</returns>
-        private static GpioDriver GetBestDriverForBoard()
+        public static GpioDriver GetBestDriverForBoard()
```
Comment (joperezr) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: Why do we want to change this to public? The main idea here is that you either want to get the driver automatic selection done for you which the default constructor is already doing, or you want to pass it in yourself, so I don't see why we would ever want to make this API public

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs [hunk 11]
```diff
@@ -28,19 +28,37 @@ public GpioController(PinNumberingScheme numberingScheme)
...
         {
             string[] cpuInfoLines = File.ReadAllLines(CpuInfoPath);
-            Regex regex = new Regex(@"Hardware\s*:\s*(.*)");
-            foreach (string cpuInfoLine in cpuInfoLines)
+            Regex regexHw = new Regex(@"Hardware\s*:\s*(.*)");
+            Regex regexRev = new Regex(@"Revision\s*:\s*(.*)");
```
Comment (joperezr) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: We are trying to move away from using Regex, which is not performant and can sometimes be unsafe to use. Is there a way to get the fact that we are running on a Pi4 via a different mechanism?

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Windows.cs [hunk 12]
```diff
@@ -34,7 +34,7 @@ public GpioController(PinNumberingScheme numberingScheme)
         ///     The GpioController could use reflection to find all GpioDriver-derived classes and call this
         ///     static method to determine if the driver considers itself to be the best match for the environment.
         /// </remarks>
-        private static GpioDriver GetBestDriverForBoard()
+        public static GpioDriver GetBestDriverForBoard()
```
Comment (joperezr) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.Windows.cs]: We don't want to make this public, so please revert back.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs [hunk 13]
```diff
@@ -12,7 +12,7 @@
 
 namespace System.Device.Gpio.Drivers
 {
-    public unsafe partial class RaspberryPi3Driver : GpioDriver
+    public unsafe partial class GenericRaspberryPiDriver : GpioDriver
```
Comment (joperezr) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: This will be an implementation detail so we shouldn't make this class public. Is it not possible to simply have the RaspberryPi4Driver to inherit from RaspberryPi3Driver and override the desired functionality?

File: src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Device.Gpio.Drivers;
+using System.Threading;
+using Xunit;
+
+namespace System.Device.Gpio.Tests
+{
+    public class RaspberryPi4DriverTests : GpioControllerTestBase
```
Comment (joperezr) [inline src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs]: This won't work until we actually have physical Raspberry Pi 4 devices on the lab in a queue. We will have to figure out how to test this before merging.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 11 above)
Comment (joperezr) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: see @Frankenslag comment above. He has been working on getting new hardware detection logic which is moving away from Regex.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 10 above)
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: what about making driver which extends/augments driver?

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs (hunk 13 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: I had it that way in the beginning, but @krwq suggested to change it this way.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 11 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: Yes, I know. I'm actually waiting for that branch to be merged so this can be combined. However, I think that we need a more sophisticated concept for detecting compatibility between drivers and hardware.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 10 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: Or (as in the test) check whether the current driver is the suggested one?

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs (hunk 6 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi4Driver.cs]: Yes, of course. Fixed.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 7 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: That was a simple typo in the comment. I don't understand your second question.

File: src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs (hunk 14 above)
Comment (krwq) [reply src/System.Device.Gpio.Tests/RaspberryPi4DriverTests.Linux.cs]: I got one Pi4 (1GB) but I haven't had time to set it up yet

File: src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs (hunk 9 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs]: I do not know what happens when you use that register on an older Pi, probably nothing you want. 
I did not change the register map start detection logic, so it uses /dev/gpiomem by default (which works). When forcing it to use /dev/mem, it seems to work as well (at least when sudo'ing).

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs (hunk 13 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: I agree on the visibility, though.

File: src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs (hunk 9 above)
Comment (joperezr) [reply src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs]: It is definitely something we will want to check as we don't want to map stuff to memory which doesn't exist. In general I would probably propose having another registerView for Pi4

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 10 above)
Comment (joperezr) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: For both of those cases, I believe we should have user-defined logic for calculating that. This is an API which by definition will have breaking changes for every new board we add, so we definitely don't want to make it public and take breaking change after breaking change. Keeping it private and changing the logic is ok, and for augmenting driver we should define a different function, and for tests we might consider using InternalsVisibleTo, which is ugly, but better than exposing this API.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs (hunk 13 above)
Comment (joperezr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: This is kind of why I suggested having an API proposal for review first, so that the design is decided before implementation so that we can first have these discussions.

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs (hunk 7 above)
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/GpioController.Linux.cs]: @pgrawehr I didn't notice `a`, `b` and `c` changing and thought those are three the same numbers

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs (hunk 13 above)
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: @pgrawehr I don't recall suggesting making `GenericRaspberryPiDriver` public, only to make it a base class

File: src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs (hunk 9 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/Interop/Unix/Libc/Interop.mmap.cs]: Yes, agree. I finally found the Pi3 register map specification: https://www.raspberrypi.org/app/uploads/2012/02/BCM2835-ARM-Peripherals.pdf (pages 89ff). This ends at address B0. 
The corresponding specification document for the BCM2711 still was not published and I took the addresses for the additional registers from https://github.com/joan2937/pigpio/blob/master/pigpio.c (at the top of the file).

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs (hunk 13 above)
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: I can make it abstract, but not internal. It's not possible to have a public class derive from an internal class.

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: Note: The current pullup/pulldown change code for the Pi3 should be moved to the Pi3 class. It doesn't seem to hurt using that on the Pi4, but its a waste of time, since that requires some delays to be introduced.

Comment (Frankenslag) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/GenericRaspberryPiDriver.Linux.cs]: @pgrawehr @krwq @joperezr 

Perhaps we should just fix the bug and at the moment I don't believe that this should involve anything more than something like the pseudocode below in the raspberrypi3 Linux driver

```
if is raspberrypi4 then
 use new pi 4 code for SetInputPullMode
else
 use existing code for SetInputPullMode
```

I think I have suggested that the test for pi4 could be based on the peripheral base address which could be cached during initialization.

**Comment by krwq** [comment]
@pgrawehr there are some merge conflicts here (probably with your other PR)

**Comment by pgrawehr** [comment]
@krwq Yes, that was to be expected. I don't know how to continue here, actually. We didn't quite get to a conclusion on #809, which I originally opened for a design discussion on this, but meanwhile went into a much broader discussion related to the "board" design topic. 

So I think until we come to a conclusion on that issue, and the result involves creating separate classes for Pi3 and Pi4, we can close this PR. The main problem is solved for now.

**Comment by pgrawehr** [comment]
Closing as per previous comment (for now)

**Comment by dontspeak** [comment]
Hello
why is this issue closed? the InputPullUp and -PullDown Resistors do not work in any way on a RP4. At least not in current version 1.0.0. Also, a `RaspberryPi4Driver` does not exist yet.
Regards Vasilios.

**Comment by joperezr** [comment]
This has been fixed in master branch, but we don't have a package in NuGet with this so it will ship with vNext. We plan on shipping some preview packages for vNext to NuGet in the next couple of weeks, but if you need this in the meantime you can just add our nightly feed to your project and restore the latest package. To do this, just add afile next to your project called NuGet.config with the following contents:
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="dotnet5" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json" />
  </packageSources>
</configuration>
```

The latest version as of today, is `1.1.0-prerelease.20078.2`

**Comment by dontspeak** [comment]
I did that and its working now.
Thank you so much!

**Comment by krwq** [comment]
@joperezr this nuget feed is different than the one on the README.md, should it be updated?

**Comment by joperezr** [comment]
Yes it should. I'll take care of that 😄
